editor/filesystem_dock_rename_dialog.cpp

```diff
@@ -0,0 +1,595 @@
+	RegEx regex(p_pattern);
+
+	return regex.sub(p_subject, p_replacement, true);
+}
+
+String FileSystemRenameDialog::_postprocess(const String &p_subject) {
+	int style_id = opt_style->get_selected();
+
+	String result = p_subject;
+
+	if (style_id == 1) {
+		// PascalCase to snake_case
+
+		result = result.to_snake_case();
+		result = _regex("_+", result, "_");
+
+	} else if (style_id == 2) {
+		// snake_case to PascalCase
+
+		RegEx pattern("_+(.?)");
+		Array matches = pattern.search_all(result);
+
+		// The name `_` would become empty; ignore it.
+		if (matches.size() && result != "_") {
+			String buffer;
+			int start = 0;
+			int end = 0;
+			for (int i = 0; i < matches.size(); ++i) {
+				start = ((Ref<RegExMatch>)matches[i])->get_start(1);
+				buffer += result.substr(end, start - end - 1);
+				buffer += result.substr(start, 1).to_upper();
+				end = start + 1;
+			}
+			buffer += result.substr(end, result.size() - (end + 1));
+			result = buffer.to_pascal_case();
+		}
+	}
+
+	int case_id = opt_case->get_selected();
+
+	if (case_id == 1) {
+		// To Lowercase
+		result = result.to_lower();
+	} else if (case_id == 2) {
+		// To Uppercase
+		result = result.to_upper();
+	}
+
+	return result;
+}
+
+void FileSystemRenameDialog::_iterate_files(const Vector<String> &p_selection, int *p_counter) {
+	for (int i = 0; i != p_selection.size(); ++i) {
+		String new_name = _apply_rename(p_selection[i], *p_counter);
+
+		if (p_selection[i].get_file() != new_name) {
+			Pair<String, String> rename_item;
+			rename_item.first = p_selection[i];
+			rename_item.second = new_name;
+			to_rename.push_back(rename_item);
+		}
+
+		*p_counter += spn_count_step->get_value();
+	}
+}
+
+void FileSystemRenameDialog::set_selected_files(const Vector<String> &p_paths, const bool &p_is_tree) {
+	selected_files = p_paths;
+	is_tree = p_is_tree;
+}
+
+void FileSystemRenameDialog::rename() {
+	global_count = spn_count_start->get_value();
+	to_rename.clear();
+
+	_iterate_files(selected_files, &global_count);
+
+	if (!to_rename.is_empty()) {
+		float delay = 1.0f;
+		EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+		undo_redo->create_action(TTR("Batch Rename"));
+
+		for (int i = 0; i != to_rename.size(); ++i) {
+			const String &filepath = to_rename[i].first;
+			const String &new_filename = to_rename[i].second;
+
+			const String &old_filename = to_rename[i].first.get_file();
+			const String &undo_filepath = to_rename[i].first.replace(old_filename, new_filename);
+			undo_redo->add_do_method(file_system_dock, "_batch_rename", filepath, new_filename, is_tree);
+			undo_redo->add_undo_method(file_system_dock, "_batch_rename", undo_filepath, old_filename, is_tree);
+		}
+
+		undo_redo->commit_action();
+
+		Ref<SceneTreeTimer> timmer = get_tree()->create_timer(delay);
+		timmer->connect("timeout", callable_mp(this, &FileSystemRenameDialog::_rescan));
```

Comment:
There is a bug with multithreading when EditorFileSystem is performing rescanning. It only occurs when there are more than 2 files being renamed.

Apparently, this code does not intend to fix it; instead, it is a workaround for that issue. When that issue is fixed, this code can be removed.